doc/design/design_model_zoo.md

```diff
@@ -0,0 +1,99 @@
+# Model Zoo
+
+SQLFlow model zoo is a place to store model definitions, pre-trained model weights and model documentations. You can directly train, predict, analyze using one of the models using SQLFlow, or you can do model fine-tune, transfer learning to use the model to fit your own dataset.
+
+SQLFlow should support below features to support common cases in machine learning:
+
+1. Host model defination and pre-trained weights in `sqlflow.org`. e.g. `sqlflow.org/modelzoo/iris_dnn_128x32` points to a directory containing a model defination of `DNNClassifier` with 128, 32 hidden layers and pre-trained weights using the iris dataset.
+1. Download pre-trained model to predict a dataset:
+   ```sql
+   SELECT * FROM iris.predict_samples
+   PREDICT predict_result.class
+   USING sqlflow.org/modelzoo/iris_dnn_128x32;
+   ```
+1. Train a new model using the model defination:
+   ```sql
+   SELECT * FROM iris.new_iris_train
+   TRAIN sqlflow.org/modelzoo/iris_dnn_128x32
+   INTO modeldb.my_iris_dnn_model;
+   ```
+1. Transfer learning to fit a new dataset with the same features:
+   ```sql
+   SELECT * FROM iris.transfer_iris_samples
+   TRAIN sqlflow.org/modelzoo/iris_dnn_128x32
+   USING sqlflow.org/modelzoo/iris_dnn_128x32
+   INTO modeldb.my_iris_dnn_model;
+   ```
+1. Fine-tune the pre-trained model:
+   ```sql
+   SELECT * FROM iris.train
+   TRAIN sqlflow.org/modelzoo/iris_dnn_128x32
+   WITH model.learning_rate=0.001, model.learning_rate_decay="cosine_decay" ...
+   INTO modeldb.my_iris_dnn_model_fine_tune;
+   ```
+
+## The Model Zoo Hosting Service
+
+The model zoo hosting service is a file service that can be accessed from the internet.
+It serves all available models and corresponding pre-trained weight. We may need to use
+a [CDN](https://en.wikipedia.org/wiki/Content_delivery_network) service if the traffic
+becomes large.
+
+Each model is saved on the server under a unique directory like: `sqlflow.org/modelzoo/iris_dnn_128x32`.
+The directory name is responsible to explain the model's type, network structure and which dataset is
+used to train the pre-trained weights. You can access `sqlflow.org/modelzoo/iris_dnn_128x32/README.md`
+from the browser to get the model's full documentation. All models under `sqlflow.org/modelzoo` are
+developed under `https://github.com/sql-machine-learning/models` weights is only stored under
+`sqlflow.org` but not under github.
+
+The content of the directory should be like:
+
+```
+iris_dnn_128x32
+   - README.md     # model documents.
+   - model_def.py  # python file of model definition (Tensorflow Estimator model or Keras model).
```

Comment:
This is just a single file though. If we want to reuse components such as losses/evaluation_metrics/etc., we should allow a Python module instead and within this Python module users can reuse and re-import from other existing modules.

Reply:
We can not permit one model to import stuff from other models because not all models are publicly available, some models is shared to part of the users for access.

Each model should be independent to other models in the model zoo, so, if the model want to reuse some code, you can:

- for public accessible deps, put it in `requirements.txt`
- for local deps, put it in some other `.py` files under this directory.

Reply:
So there cannot be any commonly shared utility modules used for different directories? If the model zoo gets larger and larger, there will be a lot of duplicate code and when it comes to migration, it would be difficult.

Reply:
If there are modules or utilities that common, it should be in Keras or TensorFlow, since it's common among all model implementations, or it is just the model specific. Contribute to Keras or TensorFlow if these utilities are really common among models.